SimTKcommon/tests/StateTest.cpp

```diff
@@ -98,81 +97,258 @@ using namespace SimTK;
+               == Real(2));
+
+    // Allocate at Topology stage a cache entry that depends on Model stage 
+    // and is guaranteed to be valid at Time stage. In between (at Model or 
+    // Instance stage) it *may* be valid if explicitly marked so.
+    const CacheEntryIndex cx0TopoModel = s.allocateCacheEntry(Sub0, 
         Stage::Model, Stage::Time, new Value<int>(41));
+    SimTK_TEST(s.getCacheEntryAllocationStage(Sub0,cx0TopoModel)
+               == Stage::Topology);
 
-    // "realize" Topology stage
-    s.advanceSubsystemToStage(Sub0, Stage::Topology);
-    s.advanceSubsystemToStage(Sub1, Stage::Topology);
-    s.advanceSystemToStage(Stage::Topology);
+    // Here is a cache entry allocated at Topology stage, with depends-on
+    // Velocity and no good-by guarantee.
+    const CacheEntryIndex cx0TopoVelocity = s.allocateCacheEntry(Sub0, 
+        Stage::Velocity, Stage::Infinity, new Value<char>('v'));
+
+    advanceStage(s, Stage::Topology);
 
-    // Shouldn't be able to access cache entry here because this is less than its
-    // "depends on" stage.
-    SimTK_TEST_MUST_THROW(s.getCacheEntry(Sub0, cx));
+    // Topology stage is realized.
+    //----------------------------
 
+    // Shouldn't be able to access cache entry here because this is less than
+    // its "depends on" stage.
+    SimTK_TEST_MUST_THROW(s.getCacheEntry(Sub0, cx0TopoModel));
 
+    //-------------------------
     // "realize" Model stage
-    s.advanceSubsystemToStage(Sub0, Stage::Model);
-    s.advanceSubsystemToStage(Sub1, Stage::Model);
-    s.advanceSystemToStage(Stage::Model);
 
+    // Allocate at Model stage, a Position-invalidating state variable.
+    const DiscreteVariableIndex dvx0ModelPos =
+        s.allocateDiscreteVariable(Sub0, Stage::Position, new Value<int>(31));
+
+    SimTK_TEST(s.getDiscreteVarAllocationStage(Sub0,dvx0ModelPos)
+               == Stage::Model);
+    SimTK_TEST(s.getDiscreteVarInvalidatesStage(Sub0,dvx0ModelPos)
+               == Stage::Position);
+    SimTK_TEST(Value<int>::downcast(s.getDiscreteVariable(Sub0,dvx0ModelPos))
+               == 31);
+
+    advanceStage(s, Stage::Model);
+    // Model stage is realized.
+    //----------------------------
+
+    //----------------------------
     // "realize" Instance stage
-    s.advanceSubsystemToStage(Sub0, Stage::Instance);
-    s.advanceSubsystemToStage(Sub1, Stage::Instance);
-    s.advanceSystemToStage(Stage::Instance);
 
-    // Although the cache entry *could* be valid at this point,
+    // Allocate a cache entry at Instance stage that has Time as depends-on
+    // and also has a cross-subsystem dependency on discrete variable 
+    // dvx0ModelPos.
+    const CacheEntryIndex cx1InstanceTime = 
+    s.allocateCacheEntryWithPrerequisites(Sub1, Stage::Time, Stage::Velocity,
+        false, false, true, // depends on z
+        {DiscreteVarKey(Sub0,dvx0ModelPos)}, 
+        {CacheEntryKey(Sub0,cx0TopoModel)},
```

Comment:
Just tried using an initializer for an `std::vector<std::pair<int, std::string>>` and it worked, so you could probably do `{ {Sub0, dvx0ModelPos}, {Sub0, cx0TopoModel} }`. Though what you've written is clearer I think, and the std::pair initializer makes the code somewhat cryptic.